examples/launch-param-reductions.cpp

```diff
@@ -0,0 +1,345 @@
+{
+
+  std::cout << "\n\nRAJA reductions example...\n";
+
+  // _reductions_array_init_start
+//
+// Define array length
+//
+  constexpr int N = 1000000;
+
+//
+// Allocate array data and initialize data to alternating sequence of 1, -1.
+//
+  int* a = memoryManager::allocate<int>(N);
+
+  for (int i = 0; i < N; ++i) {
+    if ( i % 2 == 0 ) {
+      a[i] = 1;
+    } else {
+      a[i] = -1;
+    }
+  }
+
+//
+// Set min and max loc values
+//
+  constexpr int minloc_ref = N / 2;
+  a[minloc_ref] = -100;
+
+  constexpr int maxloc_ref = N / 2 + 1;
+  a[maxloc_ref] = 100;
+  // _reductions_array_init_end
+
+//
+// Note: with this data initialization scheme, the following results will
+//       be observed for all reduction kernels below:
+//
+//  - the sum will be zero
+//  - the min will be -100
+//  - the max will be 100
+//  - the min loc will be N/2
+//  - the max loc will be N/2 + 1
+//
+//
+
+//
+// Define index range for iterating over a elements in all examples
+//
+  // _reductions_range_start
+  RAJA::TypedRangeSegment<int> arange(0, N);
+  // _reductions_range_end
+
+//
+// Define ValLoc Type
+//
+
+  using VALLOC_INT = RAJA::expt::ValLoc<int>;
+//----------------------------------------------------------------------------//
+
+  std::cout << "\n Running RAJA sequential reductions...\n";
+
+  // _reductions_raja_seq_start
+  using LAUNCH_POL1   = RAJA::LaunchPolicy<RAJA::seq_launch_t>;
+  using LOOP_POL1     = RAJA::LoopPolicy<RAJA::seq_exec>;
+
+  int seq_sum = 0;
+  int seq_min = std::numeric_limits<int>::max();
+  int seq_max = std::numeric_limits<int>::min();
+  VALLOC_INT seq_minloc(std::numeric_limits<int>::max(), -1);
+  VALLOC_INT seq_maxloc(std::numeric_limits<int>::min(), -1);
+
+  //RAJA::forall<EXEC_POL1>(arange,
+  RAJA::launch<LAUNCH_POL1>
+    (RAJA::LaunchParams(), "SeqReductionKernel",
+    RAJA::expt::Reduce<RAJA::operators::plus>(&seq_sum),
+    RAJA::expt::Reduce<RAJA::operators::minimum>(&seq_min),
+    RAJA::expt::Reduce<RAJA::operators::maximum>(&seq_max),
+    RAJA::expt::Reduce<RAJA::operators::minimum>(&seq_minloc),
+    RAJA::expt::Reduce<RAJA::operators::maximum>(&seq_maxloc),
+     [=] RAJA_HOST_DEVICE (RAJA::LaunchContext ctx,
+                           int &_seq_sum, int &_seq_min,
+                           int &_seq_max, VALLOC_INT &_seq_minloc,
+                           VALLOC_INT &_seq_maxloc) {
+
+      RAJA::loop<LOOP_POL1>(ctx, arange, [&] (int i) {
+
+          _seq_sum += a[i];
+
+          _seq_min = RAJA_MIN(a[i], _seq_min);
+          _seq_max = RAJA_MAX(a[i], _seq_max);
+
+          _seq_minloc = RAJA_MIN(VALLOC_INT(a[i], i), _seq_minloc);
+          _seq_maxloc = RAJA_MAX(VALLOC_INT(a[i], i), _seq_maxloc);
+          //_seq_minloc.min(a[i], i);
+          //_seq_maxloc.max(a[i], i);
+          // Note : RAJA::expt::ValLoc<T> objects provide min() and max() methods
+          //        that are equivalent to the assignments with RAJA_MIN and RAJA_MAX
+          //        above.
```

Comment:
Not for this PR, but do people think it would be a good idea to provide a type that wraps a reference and lets you do the operations supported by the reducers like min, max. I've grown accustomed to our reducers that have nice methods.

Reply:
If I understand you correctly, I think that's supported. If you uncomment lines 136 $ 137, does it do what you want?

Reply:
I understand that works for the loc reducers, but the new reducers don't have a wrapper for normal min/max reduction as far as I'm aware.

Reply:
Ah, OK. I was unaware of that. It should be simple to add and a good thing for consistency.

Reply:
I added an issue, its not required for this PR. #1582